File: stripe/src/test/java/com/stripe/android/FingerprintRequestTest.java
package com.stripe.android;

import androidx.test.core.app.ApplicationProvider;

import com.stripe.android.exception.InvalidRequestException;

import java.io.UnsupportedEncodingException;
import java.util.HashMap;
import java.util.Map;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;

@RunWith(RobolectricTestRunner.class)
public class FingerprintRequestTest {
    @Test
    public void getContentType() {
        assertEquals("application/json; charset=UTF-8",
                new FingerprintRequest(new HashMap<String, Object>(), "guid")
                        .getContentType());
    }

    @Test
    public void getHeaders() {
        final Map<String, String> headers =
                new FingerprintRequest(new HashMap<String, Object>(), "guid")
                    .getHeaders();
        assertEquals(StripeRequest.DEFAULT_USER_AGENT,
                headers.get(StripeRequest.HEADER_USER_AGENT));
        assertEquals("m=guid", headers.get("Cookie"));
    }

    @Test
    public void getOutputBytes() throws UnsupportedEncodingException, InvalidRequestException {
        final TelemetryClientUtil telemetryClientUtil =
                new TelemetryClientUtil(ApplicationProvider.getApplicationContext());
        final byte[] output = new FingerprintRequest(telemetryClientUtil.createTelemetryMap(),
                "guid")
                .getOutputBytes();
        assertTrue(output.length > 0);
    }
}

File: stripe/src/test/java/com/stripe/android/FingerprintRequestTest.kt
package com.stripe.android

import androidx.test.core.app.ApplicationProvider
import com.google.common.truth.Truth.assertThat
import com.stripe.android.networking.RequestHeadersFactory
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import java.io.ByteArrayOutputStream
import kotlin.test.Test

@RunWith(RobolectricTestRunner::class)
class FingerprintRequestTest {
    private val fingerprintRequestParamsFactory = FingerprintRequestParamsFactory(
        ApplicationProvider.getApplicationContext()
    )

    private val request =
        FingerprintRequest(
            params = fingerprintRequestParamsFactory.createParams(FINGERPRINT_DATA),
            guid = FINGERPRINT_DATA.guid
        )

    @Test
    fun contentType_shouldBeApplicationJson() {
        assertThat(request.contentType)
            .isEqualTo("application/json; charset=UTF-8")
    }

    @Test
    fun headers_shouldReturnExpectedMap() {
        assertThat(request.headers)
            .isEqualTo(
                mapOf(
                    "Cookie" to "m=${FINGERPRINT_DATA.guid}",
                    "User-Agent" to RequestHeadersFactory.getUserAgent(),
                    "Accept-Charset" to "UTF-8"
                )
            )
    }

    @Test
    fun writeBody_shouldWriteNonEmptyBytes() {
        ByteArrayOutputStream().use {
            request.writeBody(it)
            assertThat(it.size())
                .isGreaterThan(0)
        }
    }

    private companion object {
        private val FINGERPRINT_DATA = FingerprintDataFixtures.create()
    }
}
